File: problems/1218_longest_arithmetic_subsequence_of_given_difference/Solution.java
import java.util.*;

class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        Map<Integer, Integer> dp = new HashMap<>();
        int res = 1;
        for (int x : arr) {
            int val = dp.getOrDefault(x - difference, 0) + 1;
            dp.put(x, val);
            res = Math.max(res, val);
        }
        return res;
    }
}

File: problems/1218_longest_arithmetic_subsequence_of_given_difference/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int,int> dp;
        int res = 1;
        for (int x : arr) {
            dp[x] = dp[x - difference] + 1;
            res = max(res, dp[x]);
        }
        return res;
    }
};
